### DISCUSSION TITLE
Flashing figure when real-time update of plot

### DISCUSSION THREAD

**Discussion by MaBnt** [alandefreitas/matplotplusplus#92] [Q&A]
Hello all,

I am trying to have a figure updating a solution of a simple 1D linear advection equation in real-time, the only issue I have is that the whole window figure is flashing each time I update the solution instead of having only the curve updated. (backend gnuplot with wxt on windows 10)

I remember (no longer have a license...) that in Matlab a way to do it would be to cla() the axes and re-plot the updated data while introducing a pause in the while loop to make it possible to see the propagation at a reasonable speed. Here is a snippet of the code:
```
auto figureSolution = matplot::figure(false);
figureSolution->current_axes()->plot(physicalDomain, solution_,"-+");
while (t0 < tf) {
	upwindScheme(solution_.begin(), solution_.end(), flowNumber_);
	t0 += deltaT;
	figureSolution->current_axes()->clear();
	figureSolution->current_axes()->plot(physicalDomain, solution_);
	figureSolution->draw();
	std::this_thread::sleep_for(std::chrono::milliseconds(500));
} 
```

My questions are the following:
1) Is there a way to do that properly with gnuplot backend ?
EDIT: _2) If no is the limitation coming only from the backend or also matplot++ ?_ After testing solution with OpenGL as backend results are much better and I am getting a smooth real-time rendering.

Thanks,
Marc

**Comment by Inferno171** [comment]
can you tell me how to use opengl in backend ?


Thanks!!

**Reply by alandefreitas** [reply]
The gnuplot backend does blink in some cases. Some workarounds are to avoid creating complex graphs in a loop, by just altering what needs to be altered, or to not plot again unless some reasonable time has passed since the last drawing command.

The current OpenGL backend is still experimental. It only works for a few types of plots.
You can try this backend by building from source with the -BUILD_EXPERIMENTAL_OPENGL_BACKEND=ON option.

I'm currently working on a complete OpenGL backend but this is still going to take a few months.
It's probably going to become the default backend when done.

After that, creating backends for Qt, Wx, xeus-cling and even wasm should be relatively trivial.

**Reply by abdedixit** [reply]
Hi @alandefreitas , I am experiencing the same blinking issue for real-time plot updates with the latest version that I installed using homebrew. Is the OpenGL backend now complete or do I need to build the source using -BUILD_EXPERIMENTAL_OPENGL_BACKEND=ON option? Also please suggest if there is any better way of plotting real-time data with the current version of matplot++.

Best,
AB

**Comment by MlLearnerAkash** [comment]
Is there any update on this? I am still facing the issue on the same while updating the values in real time

**Reply by alandefreitas** [reply]
There's no good solution for that. This is a problem with how matplot++ needs to communicate with gnuplot. It can only be solved by another version of the library that uses a new backend (probably SDL or some library where we have more control over these details) but we don't have people to work on that.

**Comment by MlLearnerAkash** [comment]
Hi @alandefreitas , I have figured out a round-about solution for that. After saving the image into a temporary file and the reading that file using opencv for visualisation solves the problem of blinking of the image.

**Reply by alandefreitas** [reply]
Isn't that very expensive for high FPS?

**Reply by MlLearnerAkash** [reply]
Yes, writing and reading consumes time. But, it is giving more than 30FPS in local system with GnuPlot backend.
